File: app/src/main/java/yagoc/Move.java
package yagoc;

import yagoc.pieces.Piece;
import yagoc.pieces.PieceColor;

import static yagoc.pieces.PieceType.King;
import static yagoc.pieces.Pieces.none;

public class Move {
    public static final String[] FILE_NAMES = {"a", "b", "c", "d", "e", "f", "g", "h"};
    public static final String[] RANK_NAMES = {"8", "7", "6", "5", "4", "3", "2", "1"};

    private final Piece fromPiece;
    private final Square from;
    private final Square to;

    public Move(Piece fromPiece, Square from, Square to) {
        this.fromPiece = fromPiece;
        this.from = from;
        this.to = to;

        if (fromPiece.equals(none)) {
            throw new IllegalArgumentException("Cannot move nothing");
        }

        if (!this.from.exists() || !this.to.exists()) {
            throw new IllegalArgumentException("This move contains non-existing squares: " + from + " " + to);
        }
    }

    public String toString() {
        return fromPiece + " " + RANK_NAMES[from.rank()] + FILE_NAMES[from.file()] + " " + RANK_NAMES[to.rank()] + FILE_NAMES[to.file()];
    }

    /**
     * positive if going ahead, negative if going backwards
     */
    public int rankDistance() {
        return (to.rank() - from.rank()) * (fromPiece.color() == PieceColor.whiteSet ? -1 : 1);
    }

    /**
     * positive if going to the right, negative if going to the left
     */
    public int fileDistance() {
        return (to.file() - from.file()) * (fromPiece.color() == PieceColor.whiteSet ? 1 : -1);
    }

    public Square enPassantSquare() {
        return this.to().previousRank(this.fromPiece().color());
    }

    public int fileDistanceAbs() {
        return Math.abs(fileDistance());
    }

    public int rankDistanceAbs() {
        return Math.abs(rankDistance());
    }

    public boolean hasSameFile() {
        return to.file() == from.file();
    }

    public boolean hasSameRank() {
        return from.rank() == to.rank();
    }

    public boolean isCastling() {
        return fromPiece.pieceType() == King && fileDistanceAbs() == 2 && rankDistance() == 0;
    }

    public boolean isCastlingQueenside() {
        return isCastling() && this.to.file() < this.from.file();
    }

    public Piece fromPiece() {
        return fromPiece;
    }

    public Square from() {
        return from;
    }

    public Square to() {
        return to;
    }
}

File: app/src/main/kotlin/yagoc/board/Move.kt
package yagoc.board

import yagoc.pieces.Piece
import yagoc.pieces.PieceColor
import yagoc.pieces.PieceType
import yagoc.pieces.none
import java.io.Serializable
import kotlin.math.abs

class Move(private val fromPiece: Piece, private val from: Square, private val to: Square) : Serializable {
    override fun toString(): String {
        return fromPiece.toString() + " " + RANK_NAMES[from.rank()] + FILE_NAMES[from.file()] + " " + RANK_NAMES[to.rank()] + FILE_NAMES[to.file()]
    }

    /**
     * positive if going ahead, negative if going backwards
     */
    fun rankDistance(): Int {
        return (to.rank() - from.rank()) * if (fromPiece.color() == PieceColor.whiteSet) -1 else 1
    }

    /**
     * positive if going to the right, negative if going to the left
     */
    fun fileDistance(): Int {
        return (to.file() - from.file()) * if (fromPiece.color() == PieceColor.whiteSet) 1 else -1
    }

    fun enPassantSquare(): Square {
        return this.to().previousRank(fromPiece().color())
    }

    fun fileDistanceAbs(): Int {
        return abs(fileDistance())
    }

    fun rankDistanceAbs(): Int {
        return abs(rankDistance())
    }

    fun hasSameFile(): Boolean {
        return to.file() == from.file()
    }

    fun hasSameRank(): Boolean {
        return from.rank() == to.rank()
    }

    val isCastling: Boolean = fromPiece.pieceType() == PieceType.King && fileDistanceAbs() == 2 && rankDistance() == 0
    val isCastlingQueenside: Boolean = isCastling && to.file() < from.file()

    fun fromPiece(): Piece {
        return fromPiece
    }

    fun from(): Square {
        return from
    }

    fun to(): Square {
        return to
    }

    companion object {
        @JvmField
        val FILE_NAMES = arrayOf("a", "b", "c", "d", "e", "f", "g", "h")

        @JvmField
        val RANK_NAMES = arrayOf("8", "7", "6", "5", "4", "3", "2", "1")
    }

    init {
        require(fromPiece != none) { "Cannot move nothing" }
        require(from.exists() && to.exists()) { "This move contains non-existing squares: $from $to" }
    }
}